Giorgi's Web Forms to Blazor migration post (#839)
oktadev/okta-blog: _source/_posts/2021-08-06-webforms-blazor-migration.md

author: giorgi-dalakishvili
by: contractor
communities: [.net]
description: "Moving from .NET 4x Web Forms to Blazor? Here is a quick guide for that."
tags: [blazor, webforms, migration, dotnet, dotnetcore, csharp, aspnet, aspnetcore, dotnet5]
tweets:
- "Learn how migrate to #Blazor from #WebForms in a snap ->"
- "Get up to speed on moving to #Blazor from #WebForms in #dotnet"
- "Wondering how #Blazor stacks up against #WebForms? We've got you covered in this migration tutorial!"
image: blog/featured/okta-dotnet-blazor-closeup.jpg
type: conversion
---

ASP.NET Web Forms framework has been the cornerstone technology of .Net for web development since the release of .Net Framework in 2002. ASP.NET Web Forms includes a layer of abstraction for developers so that you don't need to care about HTML, JavaScript, or any other front-end technology. It provides a development flow similar to building desktop apps, a way for developers to build a web page by drag and drop, an event-driven programming model, and stateful controls thanks to the built-in state management. These features incurred lots of complexity, but it was worthwhile when web development was still in its infancy.

> Web development has evolved a lot in recent years. Browsers have become more powerful and implement various standards that allow access to user's devices. Client-side frameworks such as React and Angular simplify client-side development and have become increasingly popular. These frameworks and other benefits are available only if you can write the front-end JavaScript code.

Reviewer: ```suggestion
Web development has evolved a lot in recent years. Browsers have become more powerful and implement various standards that allow access to users' devices. Client-side frameworks such as React and Angular simplify client-side development and have become increasingly popular. These frameworks and other benefits are available only if you can write the front-end JavaScript code.
```
grammar fix: it's either "a user's devices" OR "users' devices" - i went for the plural

## Introduction to Blazor

WebAssembly and [Blazor](https://dotnet.microsoft.com/apps/aspnet/web-apps/blazor) come into play to address the challenge of managing front-end and back-end development effectively. WebAssembly is a new open standard of byte code for the Web that runs in every modern browser. Blazor is a C# based client-side UI framework that runs directly in the browser via WebAssembly. It can also run on the server-side and process user interaction over a real-time connection with the browser.

Blazor and ASP.NET Web Forms have a lot in common: They are both component-oriented, event-driven, and provide a stateful programming model. If you are an ASP.NET Web Forms developer, Blazor will feel familiar.

In this article, I will describe how to migrate a simple ASP.NET Web Forms project to Blazor.

To continue, you will need:

- A basic knowledge of C#
- [Visual Studio 2019](https://visualstudio.microsoft.com/vs/) or [Visual Studio Code](https://code.visualstudio.com/)
- [.NET Core SDK](https://dotnet.microsoft.com/download/dotnet-core)
- An [Okta Developer Account](https://developer.okta.com/) (free forever, to handle your OAuth needs)

> ## Overview of the WebForms Project

Reviewer: ```suggestion
## Overview of the Web Forms Project
```
Consistency of 2-word usage

    public int? FoodId { get; set; }
    public Food Food { get; set; }

    public int? ExerciseId { get; set; }
    public Exercise Exercise { get; set; }
}

public class Food
{
    public int Id { get; set; }
    public string Name { get; set; }
    public int Calories { get; set; }
}

public class Exercise
{
    public int Id { get; set; }
    public string Name { get; set; }
    public int Calories { get; set; }
}
```

Now, let's migrate the project to Blazor. First, I will create a new Blazor project.

## Create a Blazor Project

Fire up Visual Studio, create a new project by clicking **File>New Project**, select **Blazor App**, and click **Next**. Name the app *CaloriesTracker.Blazor* and click *Create*.

{% img blog/webforms-blazor-migration/image2.png alt:"" width:"800" %}{: .center-image }

In the next window, select **Blazor Server App**. Now, click **Create**, and Visual Studio will create a new server-side Blazor project.

{% img blog/webforms-blazor-migration/image3.png alt:"" width:"800" %}{: .center-image }

Alternatively, you can create the project by running the following command in a command shell:

```sh
dotnet new blazorserver -o CaloriesTracker.Blazor
```

> Blazor Server hosts Razor components on the server in an ASP.NET Core app and handles UI updates over a SignalR connection. The Blazor runtime sends events from browser to server and applies UI updates to the components that the server sends back. This strategy is similar to the approach used by ASP.NET AJAX `UpdatePanel` control that handles partial page updates in response to trigger events. The main difference is that Blazor Server maintains an active connection to the server, and all the UI state stays on the server instead of `ViewState`.

Reviewer: ```suggestion
Blazor Server hosts Razor components on the server in an ASP.NET Core app and handles UI updates over a SignalR connection. The Blazor runtime sends events from browser to server and applies UI updates to the components that the server sends back. This strategy is similar to the approach used by the ASP.NET AJAX `UpdatePanel` control that handles partial page updates in response to trigger events. The main difference is that Blazor Server maintains an active connection to the server, and all the UI state stays on the server instead of `ViewState`.
```
missing article for "control"

Let's see what you get when you create a new project:

> {% img blog/webforms-blazor-migration/image4.png alt:"" width:"500" %}{: .center-image }

Reviewer: ```suggestion
{% img blog/webforms-blazor-migration/image4.png alt:"screenshot of the Solution Explorer showing the CaloriesTracker.Blazor project" width:"500" %}{: .center-image }
```
adding alt-txt for the img for a11y

Since the Blazor Server is hosted in the ASP.NET Core project, `Startup.cs` and `Program.cs` serve the same purpose. Specifically, `Startup.cs` configures services and the app's request pipeline. The *Pages* folder contains `_Host.cshtml`, which hosts the Blazor app. The `razor` files in the *Pages* folder are the Blazor pages that comprise the app. Similarly, the *Shared* folder contains shared layouts such as the app's main layout and reusable components such as the navigation menu. You can think of `MainLayout.razor` as a `Site.Master` file in the WebForms project and `NavMenu.razor` as some `ascx` component.

Click **F5** in Visual Studio or run `dotnet run` to run the project. You should see:

NewProject Run

Next, I will migrate Entity Framework Core configuration and `CaloriesTrackerService` to the Blazor project.

## Migrate Entity Framework Core to Blazor

Migrating non-UI-related code is pretty straightforward. I will use SQLite in the Blazor project, too, so first, add the necessary package to the project:

```sh
dotnet add package Microsoft.EntityFrameworkCore.Design
dotnet add package Microsoft.EntityFrameworkCore.Sqlite
```

> Now, copy `Model.cs` file from the WebForms project to the *Data* folder in the Blazor project. Next, I need to create an initial migration and apply the migration so I will run:

Reviewer: ```suggestion
Now, copy the `Model.cs` file from the WebForms project to the *Data* folder in the Blazor project. Next, I need to create an initial migration and apply the migration, so I will run:
```
article was missing from "file" 
missing comma

{
    CaloriesTrackerService.AddFood(addItemModel.FoodId, DateTime.Today);
    dailyList = CaloriesTrackerService.GetDailyList(DateTime.Today);
    addItemModel.FoodId = -1;
}

private void HandleAddExerciseValidSubmit()
{
    CaloriesTrackerService.AddExercise(addItemModel.ExerciseId, DateTime.Today);
    dailyList = CaloriesTrackerService.GetDailyList(DateTime.Today);
    addItemModel.ExerciseId = -1;
}

class AddItemModel
{
    public int FoodId { get; set; }
    public int ExerciseId { get; set; }
}
```

Let's go through the markup and the code. First, check if the data has been loaded. If data hasn't loaded, display a *Loading* text in the browser.

Once the list of foods and exercises loads, the markup renders two forms: a food intake form and an exercise form. Both forms display a `select` element that shows items from a list of foods or exercises. The `select` value is bound to either the `addItemModel.FoodId` or the `addItemModel.ExerciseId` property using the `bind-Value` attribute. Both forms are bound to the `addItemModel` model, and when the form is submitted, it will invoke either the `HandleAddFoodValidSubmit` or the `HandleAddExerciseValidSubmit` method.

> The `OnInitialized` method runs when the component initializes, and I use it to load data for the food and exercise elements. If you need to perform asynchronous initialization, override the `OnInitializedAsync` method instead. After the data is loaded and `initialized` is set to true, Blazor will re-render the component. The `HandleAddFoodValidSubmit` and `HandleAddExerciseValidSubmit` methods handle the form submission and saving of the selected food or exercise items to the database. Switch to the browser, and you should see the two select boxes:

Reviewer: ```suggestion
The `OnInitialized` method runs when the component initializes, and I use it to load data for the food and exercise elements. If you need to perform asynchronous initialization, override the `OnInitializedAsync` method instead. After the data is loaded, and `initialized` is set to true, Blazor will re-render the component. The `HandleAddFoodValidSubmit` and `HandleAddExerciseValidSubmit` methods handle the form submission and saving of the selected food or exercise items to the database. Switch to the browser, and you should see the two select boxes:
```
adding comma for semantic clarity

}

private void FilterByDate()
{
    dailyList = CaloriesTrackerService.GetDailyList(changeDateModel.FilterDate);
}

private void HandleAddFoodValidSubmit()
{
    CaloriesTrackerService.AddFood(addItemModel.FoodId, changeDateModel.FilterDate);
    dailyList = CaloriesTrackerService.GetDailyList(changeDateModel.FilterDate);
    addItemModel.FoodId = -1;
}

private void HandleAddExerciseValidSubmit()
{
    CaloriesTrackerService.AddExercise(addItemModel.ExerciseId, changeDateModel.FilterDate);
    dailyList = CaloriesTrackerService.GetDailyList(changeDateModel.FilterDate);
    addItemModel.ExerciseId = -1;
}

class AddItemModel
{
    public int FoodId { get; set; }
    public int ExerciseId { get; set; }
}

class ChangeDateModel
{
    public DateTime FilterDate { get; set; }
}
```

Go back to the browser, and you should be able to track your calories:

{% img blog/webforms-blazor-migration/image6.png alt:"" width:"800" %}{: .center-image }

Now that I have finished migrating the code from WebForms to Blazor, I will secure the app so that no one except me knows how much sweets I eat 😃.

## Securing Your Blazor Application

> In this tutorial, I will use Okta to quickly secure the API in order not to have to roll my own identity provider and deal with all the tricky details. Okta supports user authentication, multi-factor authentication, social authentication, and all OpenID connect flows out of the box. These features take care of all the scenarios you might ever need.

Reviewer: ```suggestion
In this tutorial, I will use Okta to quickly secure the API in order not to have to roll my own identity provider and deal with all the tricky details. Okta supports user authentication, multi-factor authentication, social authentication, and all OpenID Connect flows out of the box. These features take care of all the scenarios you might ever need.
```
Correct capitalization of OIDC product name

I will use [Okta CLI](https://cli.okta.com/) to manage my Okta account and register a new application for this project.

> Run `okta start` if you don't have an Okta account, or run `okta login` if you already have an account.

Reviewer: ```suggestion
Run `okta start` if you don't have an Okta account, or run `okta login` if you already have one.
```
remove redundancy

Once you configure Okta CLI with your account, run `okta apps create` to create a new Okta application. I will use *CaloriesTracker* again as the name of the app.

On the *Type of Application* step, enter **1** to select **Web**, and on the next step, enter **4** for **Other**.

I will use `https://localhost:44360/authorization-code/callback` and accept the suggested URL for *Post Logout Redirect URI* for the redirect URI. When you hit enter, the Okta CLI creates a new app and saves the configuration to your home directory in `.okta.env` file:

{% img blog/webforms-blazor-migration/image7.png alt:"" width:"800" %}{: .center-image }

Next, go back to Visual Studio and add `Okta.AspNetCore` package:

```sh
dotnet add package Okta.AspNetCore
```

> Next, open the `appsettings.Development.json` file, add a new section `Okta` below the `Logging` section, and put your app's details. Copy `clientId` and `clientSecret` from the `.okta.env`:

Reviewer: ```suggestion
Next, open the `appsettings.Development.json` file, add a new section `Okta` below the `Logging` section, and put in your app's details. Copy `clientId` and `clientSecret` from the `.okta.env`:
```
Correct verb usage

```json
{
  "Logging": {
    "LogLevel": {
    "Default": "Information",
    "Microsoft": "Warning",
    "Microsoft.Hosting.Lifetime": "Information"
    }
  },
  "Okta": {
    "ClientId": "{clientId}",
    "ClientSecret": "{clientSecret}",
    "Domain": "https://{yourOktaDomain}"
  }
}
```

Now, go the `Startup` class and add this code in `ConfigureServices` method:

```cs
services.AddAuthentication(options =>
{
    options.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
    options.DefaultSignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
    options.DefaultChallengeScheme = OktaDefaults.MvcAuthenticationScheme;
})
.AddCookie()
.AddOktaMvc(new OktaMvcOptions
{
    OktaDomain = Configuration["Okta:Domain"],
    ClientId = Configuration["Okta:ClientId"],
    ClientSecret = Configuration["Okta:ClientSecret"]
});
```

Also, add the following line to the `Configure()` method after the `app.UseRouting();` line:

```cs
app.UseAuthentication();
```

> Next, let's create `login` and `logout` pages. As Blazor uses SignalR for the client and server-side communication and Okta uses OpenID protocol that relies on HTTP, you need to perform authentication outside of the Blazor app.

Reviewer: ```suggestion
Next, let's create `login` and `logout` pages. As Blazor uses SignalR for the client and server-side communication and Okta uses the OpenID protocol that relies on HTTP, you need to perform authentication outside of the Blazor app.
```
Add an article - 'the'

    await HttpContext.SignOutAsync(CookieAuthenticationDefaults.AuthenticationScheme);
}
```

This code ends the Okta session and signs out the user.

Now, let's plug these pages into the Blazor app.

Open the `App.razor` file and replace it's content with this:

```html
<Router AppAssembly="@typeof(Program).Assembly">
    <Found Context="routeData">
        <AuthorizeRouteView RouteData="@routeData" DefaultLayout="@typeof(MainLayout)">
        <Authorizing>
            <p>Authorizing please wait...</p>
        </Authorizing>
        <NotAuthorized>
            <h1>Sorry</h1>
            <p>You're not authorized to see this page. You need to log in.</p>
        </NotAuthorized>
        </AuthorizeRouteView>
    </Found>
    <NotFound>    
        <p>Sorry, there's nothing at this address.</p>    
    </NotFound>
</Router>
```

The `AuthorizeRouteView` component ensures that you can see only the pages that the user is authorized to visit.

After this, open the `CaloriesTracking.razor` component and add `@attribute [Authorize]` at the top of the code so that only authorized users can view it.

> Finally, you need a way to authenticate users. I will create a new component, `AccountControl` in the `Shared` folder and put the following code:

Reviewer: ```suggestion
Finally, you need a way to authenticate users. I will create a new component, `AccountControl` in the `Shared` folder and add in the following code:
```
Correct verb usage

```html
@inject NavigationManager NavigationManager

<AuthorizeView>
    <Authorized>
        <span>Hello, @context.User.Identity.Name!</span>
        <a href="logout">Log out</a>
    </Authorized>
    <NotAuthorized>
        <a href="login?redirectUri=@(NavigationManager.Uri)">Log in</a>
    </NotAuthorized>
</AuthorizeView>
```

The `AuthorizeView` view will either display a *Log out* link for the user or a *Log in* link for authentication. The `AuthorizeView` component exposes a `context` variable, which I use to get the information about the logged-in user.

Put the component in the `MainLayout.razor` file above the *About* link and run the app. When you click the **Calories Tracker** you will see a message that you don't have access to this page:

> {% img blog/webforms-blazor-migration/image8.png alt:"" width:"800" %}{: .center-image }

Reviewer: ```suggestion
{% img blog/webforms-blazor-migration/image8.png alt:"Screenshot of the CaloriesTracker login screen for unauthorized user view" width:"800" %}{: .center-image }
```

Click the **Log in** link, and the app will redirect you to Okta for authentication. When you log in, Okta sends you back to the app, where you'll be able to access the page that only authenticated users can see:

> {% img blog/webforms-blazor-migration/image9.png alt:"" width:"800" %}{: .center-image }

Reviewer: ```suggestion
{% img blog/webforms-blazor-migration/image9.png alt:"Screenshot of CaloriesTracker interface showing "Daily consumed and burned calories" width:"800" %}{: .center-image }
```

> {% img blog/webforms-blazor-migration/image9.png alt:"Screenshot of CaloriesTracker interface showing "Daily consumed and burned calories" width:"800" %}{: .center-image }

Reviewer 2: ```suggestion
{% img blog/webforms-blazor-migration/image9.png alt:"Screenshot of CaloriesTracker interface showing 'Daily consumed and burned calories'" width:"800" %}{: .center-image }
```

That's it! You have now finished migrating a WebForms project to Blazor and secured it with Okta.

Give Blazor a try. Get the [complete source code of the project from GitHub](https://github.com/oktadev/Okta.CaloriesTracker), and let us know what you think in the comments below.

- [Goodbye Javascript! Build an Authenticated Web App in C# with Blazor + ASP.NET Core 3.0](https://developer.okta.com/blog/2019/10/16/csharp-blazor-authentication)
- [How to Build Securely with Blazor WebAssembly (WASM)](https://developer.okta.com/blog/2020/09/30/blazor-webassembly-wasm-dotnetcore)
- [ASP.NET Core 3.0 MVC Secure Authentication](https://developer.okta.com/blog/2019/11/15/aspnet-core-3-mvc-secure-authentication)

> Make sure to follow us on [Twitter](https://twitter.com/oktadev) and subscribe to our [YouTube Channel](https://youtube.com/c/oktadev) and check out our new [Twitch](https://www.twitch.tv/oktadev) live stream so that you never miss any awesome content!

Reviewer: ```suggestion
Make sure to follow us on [Twitter](https://twitter.com/oktadev),  subscribe to our [YouTube Channel](https://youtube.com/c/oktadev), and check out our new [Twitch](https://www.twitch.tv/oktadev) live stream so that you never miss any awesome content!
```
"and" reduction in a series...
